### PR TITLE
Tide: Deduplicate checkrun results and use the best only

### PR DESCRIPTION
**Pull request by alvaroaleman**
Checkruns can be restarted, resulting in mulitiple checkruns of the same
name being reported on a given commit with potentially different
results.

This PR makes tide de-duplicate them and use the best, similiar to what
we already do for ProwJobs.

### CODE REVIEW

**Comment by matthyx** [comment]
super clean implementation!
/lgtm
/hold

**Review by stevekuznetsov — APPROVED** [review]
I'm not super familiar with how check runs show up in the UI - but if we expect a success to beat out all other statuses, we will merge even if the latest run has failed, is this a case where the UI shows a checkrun failure?

/hold

**Comment by alvaroaleman** [comment]
> I'm not super familiar with how check runs show up in the UI - but if we expect a success to beat out all other statuses, we will merge even if the latest run has failed, is this a case where the UI shows a checkrun failure?

Hypothetically this could happen, the UI from what it looks always showed the latest run. But that would require a successful run to happen, then someone trigger a re-run and then that re-run to fail. While possible, its unlikely because why would you re-run a test that succeeded?

It is the very same with prowjobs btw, tide just checks if it finds any successful prowjob for the given commit, if a prowjob first succeeds, then gets manually retriggered and fails, tide would still merge.

**Pull request merged**
